File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(items: T[], predicate: (item: T) => boolean, consumer: (item: T) => U): U | undefined {
  const first = items.find(predicate)
  if (first != undefined){
    return consumer(first)
  }
  return undefined
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = BigInt(1); ; power *= base){
    yield power
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  let meaningfulLines = 0
  const file = await open(filePath, "r") //Open the file
  for await (const line of file.readLines()) { //For each line in the file, filter for whitespace and check to see if it has meaning
    const filteredLine = line.replace(/\s+/g, "")
    if(filteredLine != "" && !filteredLine.startsWith("#")){
      meaningfulLines += 1
    }
  }
  file.close()
  return meaningfulLines
}

interface Sphere{
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch(shape.kind) {
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
    case "Box":
      return 2 * (shape.width * shape.length + shape.width * shape.depth + shape.length * shape.depth)
  }
}

export function volume(shape: Shape): number {
  switch(shape.kind) {
    case "Sphere":
      return ((4/3) * Math.PI * shape.radius ** 3)
    case "Box":
      return (shape.width * shape.length * shape.depth)
  }
}

// Write your binary search tree implementation here
export interface BinarySearchTree<T> {
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }
  insert(value: T): BinarySearchTree<T> {
      return new Node(value, new Empty(), new Empty())
  }
  contains(value: T): boolean {
      return false
  }
  inorder(): Iterable<T> {
      return []
  }
  public toString(): String{
    return "()"
  }
}

class Node<T> implements BinarySearchTree<T> {
  private value!: T
  private left!: BinarySearchTree<T>
  private right!: BinarySearchTree<T>

  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>){
    this.value = value
    this.left = left
    this.right = right
  }
  size(): number {
      return 1 + this.left.size() + this.right.size()
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value && !this.contains(value)){
      return new Node(this.value, this.left.insert(value), this.right);
    } else if (!this.contains(value)){
      return new Node(this.value, this.left, this.right.insert(value));
    } else {
      return this
    }
  }

  contains(value: T): boolean {
    return this.value == value || this.left.contains(value) || this.right.contains(value);
  }

  *inorder(): Iterable<T> {
    if(this.size() == 0){
      return;
    }

    yield* this.left.inorder()

    yield this.value

    yield* this.right.inorder()
  }

  public toString(): String {
    return `(${this.left.size() != 0 ? this.left : ""}${this.value}${this.right.size() != 0 ? this.right : ""})`
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply array predicate consumer = (* Find the optional first item in the array that satisfies the predicate and then consume it*)
  match List.find_opt predicate array with
  | None -> None
  | Some x -> consumer x;;

let powers_generator base = (* Recursively generate a sequence of powers *)
  let rec generate_from power () =
    Seq.Cons (power, generate_from (power * base))
  in
  generate_from 1;;

let meaningful_line_count filename = (* Counts the number of meaningful lines in a file *)
  let meaningful_line line = (* a meaningful line is defined as a non-whitespace line that doesn't start with "#" *)
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.get trimmed 0 = '#')
  in
  let the_file = open_in filename in (* Open the file *)
  let finally () = close_in the_file in (* Make sure we close the file at the end *)
  let rec count_lines count = (* Recursively go through the lines and compare to the meaningful_line definition above *)
    try
      let line = input_line the_file in
      if meaningful_line line then
        count_lines (count + 1)
      else
          count_lines count
    with
    | End_of_file -> count (* Return count at the end of the file *)
  in
  Fun.protect ~finally (fun () -> count_lines 0);; (* Protect the function so it auto-closes and then start counting *)

type shape =  (* Shape type *)
  | Sphere of float
  | Box of float * float * float

let volume shape = (* Returns volume of a shape *)
  match shape with
  | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
  | Box (l, w, h) -> l *. w *. h;;

let surface_area shape = (* Returns surface area of a shape *)
  match shape with
  | Sphere r -> 4. *. Float.pi *. (r ** 2.)
  | Box (l, w, h) -> 2. *. ((l *. w) +. (w *. h) +. (h *. l))


type 'a binary_search_tree = (* BST of generic type 'a *)
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size bst = (* Returns size of a BST *)
  match bst with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right;;

let rec insert value bst = (* Inserts a value into a BST *)
  match bst with
  | Empty -> Node (Empty, value, Empty)
  | Node (left, nodeValue, right) ->
    if value < nodeValue then
      Node ((insert value left), nodeValue, right)
    else if value > nodeValue then
      Node (left, nodeValue, (insert value right))
    else
      Node (left, nodeValue, right)

let rec contains value bst = (* Checks to see if the given value is in the BST *)
  match bst with
  | Empty -> false
  | Node (left, nodeValue, right) -> 
    if nodeValue = value then
      true
    else if value < nodeValue then
      contains value left
    else
      contains value right;;

let rec inorder bst = (* Returns in order traversal of a BST *)
  match bst with
  | Empty -> []
  | Node (left, v, right) -> inorder left @ [v] @ inorder right;;
